typing_extensions/src_py3/typing_extensions.py

```diff
@@ -620,3 +665,269 @@ def new_type(x):
+                continue
+            annotations = getattr(base, '__annotations__', {})
+            for attr in list(base.__dict__.keys()) + list(annotations.keys()):
+                if (not attr.startswith('_abc_') and attr not in (
+                        '__abstractmethods__', '__annotations__', '__weakref__',
+                        '_is_protocol', '_is_runtime_protocol', '__dict__',
+                        '__args__', '__slots__', '_get_protocol_attrs',
+                        '__next_in_mro__', '__parameters__', '__origin__',
+                        '__orig_bases__', '__extra__', '__tree_hash__',
+                        '__doc__', '__subclasshook__', '__init__', '__new__',
+                        '__module__', '_MutableMapping__marker', '_gorg') and
+                        getattr(base, attr, object()) is not None):
+                    attrs.add(attr)
+        return attrs
+
+    @_tp_cache
+    def __getitem__(self, params):
+        # We also need to copy this from GenericMeta.__getitem__ to get
+        # special treatment of "Protocol". (Comments removed for brevity.)
+        if not isinstance(params, tuple):
+            params = (params,)
+        if not params and self._gorg is not Tuple:
+            raise TypeError(
+                "Parameter list to %s[...] cannot be empty" % self.__qualname__)
+        msg = "Parameters to generic types must be types."
+        params = tuple(_type_check(p, msg) for p in params)
+        if self in (Generic, Protocol):
+            if not all(isinstance(p, TypeVar) for p in params):
+                raise TypeError(
+                    "Parameters to %r[...] must all be type variables" % self)
+            if len(set(params)) != len(params):
+                raise TypeError(
+                    "Parameters to %r[...] must all be unique" % self)
+            tvars = params
+            args = params
+        elif self in (Tuple, Callable):
+            tvars = _type_vars(params)
+            args = params
+        elif self.__origin__ in (Generic, Protocol):
+            raise TypeError("Cannot subscript already-subscripted %s" %
+                            repr(self))
+        else:
+            _check_generic(self, params)
+            tvars = _type_vars(params)
+            args = params
+
+        prepend = (self,) if self.__origin__ is None else ()
+        return self.__class__(self.__name__,
+                              prepend + self.__bases__,
+                              _no_slots_copy(self.__dict__),
+                              tvars=tvars,
+                              args=args,
+                              origin=self,
+                              extra=self.__extra__,
+                              orig_bases=self.__orig_bases__)
+
+if NO_PROTOCOL:
+    del _ProtocolMeta
+else:
+    class Protocol(metaclass=_ProtocolMeta):
+        """Base class for protocol classes. Protocol classes are defined as::
+
+          class Proto(Protocol[T]):
```

Comment:
See my comments about the corresponding docstring in the Python 2 implementation. They are also relevant here.